File: server/Game.cpp
#include "Game.h"
#include "Command.h"
#include "Parser.h"
#include "Card.h"

int Game::GAME_COUNTER = -1;

Game::Game() {
    // ID number of the game by incrementing counter
    game_id = ++GAME_COUNTER;
    // Default starting game state
    init();
}

void Game::init() {
    black_p = nullptr;
    white_p = nullptr;
    is_active = false;
    curr_p = nullptr;
    white_to_move = true;
    winner_p = nullptr;

    vector<vector<string>> def_board
            {
                    {"bP", "bP", "bK", "bP", "bP"},
                    {"--", "--", "--", "--", "--"},
                    {"--", "--", "--", "--", "--"},
                    {"--", "--", "--", "--", "--"},
                    {"wP", "wP", "wK", "wP", "wP"}
            };
    board = def_board;
    five_cards = Card::pick_five_cards();

    piece_moved = "";
    piece_captured = "";
}

void Game::enter_game(Player * player){
    // Black player slot available
    if (black_p == nullptr){
        black_p = player;

        string send_text;
        send_text.append(Command::name(Cmd::WAITING))
                .append(1, Parser::SPL)
                .append(player->user)
                .append(1, Parser::END);
        send(player->sock, send_text.data(), send_text.size(), 0);
        Parser::send_log(player->sock, send_text);

        cout << "Player1 (Black) " << player->user << " has entered lobby!" << endl;
    }
    // White player slot available
    else if(white_p == nullptr){
        white_p = player;

        string send_text;
        send_text.append(Command::name(Cmd::WAITING))
                .append(1, Parser::SPL)
                .append(player->user)
                .append(1, Parser::END);
        send(player->sock, send_text.data(), send_text.size(), 0);
        Parser::send_log(player->sock, send_text);

        cout << "Player2 (White) " << player->user << " has entered lobby!" << endl;

        // GAME started
        start_game();
    }
}

void Game::start_game(){
    string send_text;
    send_text.append(Command::name(Cmd::START))
            .append(1, Parser::SPL)
            .append(black_p->user)
            .append(1, Parser::SPL)
            .append(white_p->user)
            .append(1, Parser::SPL);

    // Pick 5 random cards first BLACK SPARE WHITE
    int last_index = five_cards.size() - 1;
    for (int i = 0; i < last_index; ++i){
        send_text.append(five_cards[i])
            .append(1, Parser::SPL);
    }
    send_text.append(five_cards[last_index])
        .append(1, Parser::END);

    // Send to both TWO players START
    send(black_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(black_p->sock, send_text);
    black_p->state = State_Machine::allowed_transition(black_p->state, Event::EV_PLAY);

    send(white_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(white_p->sock, send_text);
    white_p->state = State_Machine::allowed_transition(white_p->state, Event::EV_PLAY);

    is_active = true;
    white_to_move = true;
    curr_p = white_p;
    cout << "Game started with id: " << game_id << endl;
}

bool Game::check_move(const string& card, int st_row, int st_col, int end_row, int end_col){
    vector<tuple<int,int>> positions;
    int new_r, new_c, fst_card, snd_card;
    string piece_color;
    bool inverted;

    // Black player up there
    if (!white_to_move){
        inverted = true;
        fst_card = 0;
        snd_card = 1;
        piece_color = "b";
    }
    // White player down there
    else{
        inverted = false;
        fst_card = 3;
        snd_card = 4;
        piece_color = "w";
    }

    // Card is valid
    if (Card::is_valid_card(card)){
        positions = Card::get_positions(card, inverted);
        // Using only black cards or white cards
        if(card == five_cards.at(fst_card) or card == five_cards.at(snd_card)){
            // Not outside of playing field
            if(end_row >= 0 and end_row <= 4 and end_col >= 0 and end_col <= 4){
                // Moving with our figures and not to friendly figures
                if (board[st_row][st_col].substr(0,1) == piece_color
                and board[end_row][end_col].substr(0,1) != piece_color){
                    for (tuple<int, int> vec : positions) {
                        // Transform x,y system to board system
                        new_r = st_row - get<1>(vec);
                        new_c = st_col + get<0>(vec);
                        // Card and given position is right
                        if (new_r == end_row and new_c == end_col) {
                            piece_moved = board[st_row][st_col];
                            piece_captured = board[end_row][end_col];
                            board[st_row][st_col] = "--";
                            board[end_row][end_col] = piece_moved;
                            cout << "Move was OK!" << endl;
                            return true;
                        }
                    }
                }
            }
        }
    }

    cout << "Move was invalid for some reason!" << endl;
    return false;
}

void Game::move_was_made(const string& card, int st_row, int st_col, int end_row, int end_col) {
    string send_text;
    send_text.append(Command::name(Cmd::MOVE_WAS_MADE))
            .append(1, Parser::SPL)
            .append(card)
            .append(1, Parser::SPL)
            .append(to_string(st_row))
            .append(1, Parser::SPL)
            .append(to_string(st_col))
            .append(1, Parser::SPL)
            .append(to_string(end_row))
            .append(1, Parser::SPL)
            .append(to_string(end_col))
            .append(1, Parser::END);

    send(black_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(black_p->sock, send_text);
    send(white_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(white_p->sock, send_text);
}

bool Game::valid_pass(string card) {
    vector<vector<tuple<int, int>>> pos_card_1_2;
    int new_r, new_c , fst_card, snd_card;
    string piece_color, curr_piece;
    bool inverted;

    // Iterate trough black cards
    if(!white_to_move){
        inverted = true;
        pos_card_1_2.push_back(Card::get_positions(five_cards[0], inverted));
        pos_card_1_2.push_back(Card::get_positions(five_cards[1], inverted));
        fst_card = 0;
        snd_card = 1;
        piece_color = "b";
    }
    // Iterate trough white cards
    else{
        inverted = false;
        pos_card_1_2.push_back(Card::get_positions(five_cards[3], inverted));
        pos_card_1_2.push_back(Card::get_positions(five_cards[4], inverted));
        fst_card = 3;
        snd_card = 4;
        piece_color = "w";
    }


    for (int row = 0; row < board.size(); ++row){
        for (int col = 0; col < board[row].size(); ++col){
            curr_piece = board[row][col];
            // Moving with our figures
            if (curr_piece.substr(0, 1) == piece_color){
                // Both cards
                for (const vector<tuple<int,int>>& pos : pos_card_1_2){
                    // Check for all possibilities
                    for (tuple<int, int> vec : pos) {
                        // Transform x,y system to board system
                        new_r = row - get<1>(vec);
                        new_c = col + get<0>(vec);
                        // Not outside of playing field
                        if(new_r >= 0 and new_r <= 4 and new_c >= 0 and new_c <= 4) {
                            // Not to friendly figures
                            if (board[new_r][new_c].substr(0,1) != piece_color){
                                cout << "Pass was invalid for some reason!" << endl;
                                return false;
                            }
                        }
                    }
                }
            }
        }
    }

    // Cant move so card will be swapped
    // Card is valid
    if (Card::is_valid_card(card)) {
        // Using only black cards or white cards
        if (card == five_cards.at(fst_card) or card == five_cards.at(snd_card)) {
            cout << "Pass was OK!" << endl;
            return true;
        }
    }
    // Given card does not exist
    return false;
}

void Game::pass_was_made(const string& card) {
    string send_text;
    send_text.append(Command::name(Cmd::PASS_WAS_MADE))
            .append(1, Parser::SPL)
            .append(card)
            .append(1, Parser::END);

    send(black_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(black_p->sock, send_text);
    send(white_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(white_p->sock, send_text);
}

void Game::invalid_move(string message, int fd) {
    string send_text;
    send_text.append(Command::name(Cmd::INVALID_MOVE))
            .append(1, Parser::SPL)
            .append(message)
            .append(1, Parser::END);
    send(fd, send_text.data(), send_text.size(), 0);
    Parser::send_log(fd, send_text);
}

void Game::shuffle_cards(const string& card) {
    // Black player and White Player
    string tmp;
    for (int i = 0; i < five_cards.size(); ++i){
        if (i == 2){
            continue;
        }
        else if (five_cards[i] == card){
            tmp = five_cards[i];
            five_cards[i] = five_cards[2];
            five_cards[2] = tmp;
            cout << "Cards swapped: " << tmp << " -> " << five_cards[i] << endl;
            break;
        }
    }
}

void Game::switch_players() {
    white_to_move = !white_to_move;

    if (curr_p == black_p){
        curr_p = white_p;
    }
    else{
        curr_p = black_p;
    }
    cout << "Players switched!" << endl;
}

bool Game::is_end() {
    // Checkmate || Enemy base for white || Enemy base for black
    if (piece_captured.substr(1, 1) == "K" or board[0][2] == "wK" or board[4][2] == "bK"){
        winner_p = curr_p;
        cout << "Game over encountered!" << endl;
        return true;
    }
    cout << "Not game over yet!" << endl;
    return false;
}

void Game::game_over() {
    string send_text;
    send_text.append(Command::name(Cmd::GAME_OVER))
            .append(1, Parser::SPL)
            .append(winner_p->user)
            .append(1, Parser::END);
    send(white_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(white_p->sock, send_text);
    send(black_p->sock, send_text.data(), send_text.size(), 0);
    Parser::send_log(black_p->sock, send_text);
    cout << "End of the game" << endl;

    // Rematch
    int old_sock = black_p->sock;
    black_p->init();
    black_p->sock = old_sock;

    old_sock = white_p->sock;
    white_p->init();
    white_p->sock = old_sock;
}

File: FINAL/client/onitama/game.py
import engine
import parser
from parser import Cmd
from components import *
from network import Network

import pygame


# Resolution for main game window
WIDTH_PLAY = 1024
HEIGHT_PLAY = 512
# Dimension of board (how many squares)
DIMENSION = 5
# Square size
# (x, y)
SQ_SIZE = HEIGHT_PLAY // DIMENSION

# Temp variable for Card size
CARD_COEF = 1 / 125
# Card size (x, y)
CARD_SIZE = (CARD_COEF * (300 * SQ_SIZE), CARD_COEF * (174 * SQ_SIZE))

# Card positions
# (x, y)
CARD_POS = [
    # Black player's cards
    (WIDTH_PLAY // 2, 1),
    (WIDTH_PLAY - CARD_SIZE[0], 1),
    # Spare card
    (0.55 * CARD_SIZE[0] + WIDTH_PLAY // 2, HEIGHT_PLAY // 2 - CARD_SIZE[1] // 2),
    # White player's cards
    (WIDTH_PLAY // 2, HEIGHT_PLAY - CARD_SIZE[1]),
    (WIDTH_PLAY - CARD_SIZE[0], HEIGHT_PLAY - CARD_SIZE[1]),
]

# Temp variable for Label size
LABEL_COEF = 0.13
# Label positions (x, y)
# Position relative to B1 card
BLACK_NAME_POS = (CARD_POS[0][0], CARD_POS[0][1] + CARD_SIZE[1])
BLACK_STATUS_POS = (BLACK_NAME_POS[0] , BLACK_NAME_POS[1] + CARD_SIZE[1] * LABEL_COEF)
# Relative to spare card
SPARE_CARD_POS = (CARD_POS[2][0] + CARD_SIZE[0] * 2*LABEL_COEF, CARD_POS[2][1] - CARD_SIZE[1] * LABEL_COEF)
# Position relative to W1 card
WHITE_NAME_POS = (CARD_POS[3][0], CARD_POS[3][1] - CARD_SIZE[1] * LABEL_COEF)
WHITE_STATUS_POS = (WHITE_NAME_POS[0], WHITE_NAME_POS[1] - CARD_SIZE[1] * LABEL_COEF)

# Others
MAX_FPS = 15
PIECE_IMAGES = {}
CARD_IMAGES = {}


"""
Handle user input and update graphics. Also react to messages from server and update game stats accordingly.
"""
def play(net: Network, data, username, rec_data):
    # Init pygame library
    pygame.init()
    my_font = pygame.font.SysFont(FONT, 18)
    clock = pygame.time.Clock()
    screen = pygame.display.set_mode((WIDTH_PLAY, HEIGHT_PLAY))

    # Change icon and title
    program_icon = pygame.image.load('pieces/bN.png')
    pygame.display.set_icon(program_icon)
    pygame.display.set_caption('Onitama')

    # Game state
    gs = engine.GameState(net, data, username)
    if rec_data is not None:
        gs.reconnect(rec_data)

    # Only once before loop
    load_images(gs)

    # Nullable variables
    global card_picked, valid_moves, sq_selected, player_clicks
    # No square selected, last click of user (row, col)
    sq_selected = ()
    # All the player clicks two tuple [(4,5), (2,2)]
    player_clicks = []
    # Valid moves
    valid_moves = []
    # If user selected card
    card_picked = None

    rematch = False
    running = True
    while running:
        # True if player has no option to play, so he gives back one card by clicking twice on it and pass turn
        pass_turn = gs.nowhere_to_go()
        gs.broken_conn = net.broken_connection()

        # Event in pygame
        for e in pygame.event.get():
            if e.type == pygame.QUIT:
                running = False

            # Mouse handler Only current player can play
            if e.type == pygame.MOUSEBUTTONDOWN \
                    and gs.win_white is None \
                    and gs.curr_p == gs.player_name \
                    and not gs.black_disconnected \
                    and not gs.white_disconnected \
                    and not gs.broken_conn:
                # x,y loc of mouse
                loc = pygame.mouse.get_pos()

                # Second clock Clicking on the board
                if card_picked is not None and (loc[0] <= WIDTH_PLAY // 2 and loc[1] <= HEIGHT_PLAY):
                    col = loc[0] // SQ_SIZE
                    row = loc[1] // SQ_SIZE

                    # User clicked at the same square twice
                    if sq_selected == (row, col):
                        # Reset user clicks
                        sq_selected = ()
                        player_clicks = []
                    else:
                        # Append fst and snd click
                        sq_selected = (row, col)
                        player_clicks.append(sq_selected)

                    # Second click
                    if len(player_clicks) == 2:
                        move = engine.Move(player_clicks[0], player_clicks[1], gs.board)
                        if move in valid_moves:
                            # Send MAKE_MOVE to server -> Reply will be collected in select in next iteration
                            make_move = parser.prepare_make_move(
                                str(card_picked),
                                move.start_row,
                                move.start_col,
                                move.end_row,
                                move.end_col)
                            net.send_data(make_move)
                        clear_selection()
                # Second click -> Confirm to exchange that card nad pass turn
                elif card_picked is not None and pass_turn:
                    for i, card in enumerate(gs.selected_cards):
                        if gs.white_to_move and (i == 0 or i == 1):
                            continue
                        elif i == 2:
                            continue
                        elif not gs.white_to_move and (i == 3 or i == 4):
                            continue

                        x, y, w, h = (CARD_POS[i][0], CARD_POS[i][1], CARD_SIZE[0], CARD_SIZE[1])
                        if x < loc[0] < x + w and y < loc[1] < y + h:
                            if card_picked == card:
                                make_pass = parser.prepare_make_pass(card)
                                net.send_data(make_pass)
                                print("Pass was made with card: " + card)
                                break
                    clear_selection()
                # First click -> Clicking cards on the right side
                else:
                    clear_selection()
                    for i, card in enumerate(gs.selected_cards):
                        if gs.white_to_move and (i == 0 or i == 1):
                            continue
                        elif i == 2:
                            continue
                        elif not gs.white_to_move and (i == 3 or i == 4):
                            continue

                        # Selecting one card
                        x, y, w, h = (CARD_POS[i][0], CARD_POS[i][1], CARD_SIZE[0], CARD_SIZE[1])
                        if x < loc[0] < x + w and y < loc[1] < y + h:
                            valid_moves = gs.get_valid_moves(card)
                            card_picked = card
                            print("Selected: " + card)
                            break

        # Incoming message from server
        server_rx_buffer = []
        buff = net.network_data_arrived(server_rx_buffer)
        if buff is None:
            running = False
            rematch = False
        else:
            for record in server_rx_buffer:
                data = parser.parse(record)
                cmd = data[0]

                if cmd == Cmd.PING.value:
                    if data[1] == Cmd.PING.value:
                        net.allow_ping()
                    else:
                        running = not net.wrong_attempt()

                elif cmd == Cmd.MOVE_WAS_MADE.value:
                    try:
                        print(f"Move was accepted by server and made in client!")
                        if len(data) != 6:
                            raise Exception("Server MOVE_WAS_MADE invalid!")

                        # Move OK -> switch cards and move piece
                        the_card = data[1]
                        start_tup = (int(data[2]), int(data[3]))
                        end_tup = (int(data[4]), int(data[5]))
                        the_move = engine.Move(start_tup, end_tup, gs.board)

                        # Make the move on the board
                        gs.make_move(the_move)
                        if gs.move_log:
                            animate_move(gs.move_log[-1], screen, gs.board, clock)

                        # Shuffle cards and swap players
                        gs.shuffle_cards(the_card)
                        gs.switch_players()

                    except Exception as e:
                        print(e)
                        running = not net.wrong_attempt()

                elif cmd == Cmd.PASS_WAS_MADE.value:
                    if len(data) == 2:
                        the_card = data[1]
                        print(f"Pass was accepted by server and made in client with card {the_card}!")
                        gs.shuffle_cards(the_card)
                        gs.switch_players()
                    else:
                        running = not net.wrong_attempt()

                elif cmd == Cmd.INVALID_MOVE.value:
                    if len(data) == 2:
                        print(Cmd.INVALID_MOVE.value, f"{data[1]}")
                    else:
                        running = not net.wrong_attempt()

                elif cmd == Cmd.GAME_OVER.value:
                    if len(data) == 2:
                        winner_name = data[1]
                        print(f"Game over encountered, player {winner_name} has won!")
                        gs.game_over(winner_name)
                    else:
                        running = not net.wrong_attempt()
                    rematch = True

                elif cmd == Cmd.RECONNECT.value:
                    if len(data) == 2 and data[1] == gs.enemy_name:
                        if gs.enemy_name == gs.white_name:
                            gs.white_disconnected = False
                        elif gs.enemy_name == gs.black_name:
                            gs.black_disconnected = False
                    else:
                        print("ERR: RECONNECT in Game not possible!")
                        running = not net.wrong_attempt()

                elif cmd == Cmd.DISCONNECT.value:
                    if len(data) == 2 and data[1] == gs.enemy_name:
                        if gs.enemy_name == gs.white_name:
                            gs.white_disconnected = True
                        elif gs.enemy_name == gs.black_name:
                            gs.black_disconnected = True
                    else:
                        print("ERR: DISCONNECT in Game not possible!")
                        running = not net.wrong_attempt()

                elif cmd == "WRONG_DATA":
                    print("ERR: WRONG_DATA", f"Data are not parsable!")
                    running = not net.wrong_attempt()

                else:
                    print(f"ERR: Unknown message in Game!")
                    running = not net.wrong_attempt()

        draw_game_state(screen, gs, valid_moves, sq_selected, card_picked, my_font)
        clock.tick(MAX_FPS)
        pygame.display.flip()

    print("Game finished - quiting pygame!")
    pygame.quit()
    return rematch


"""
Clear player selection so nothing is chosen
"""
def clear_selection():
    global card_picked, valid_moves, sq_selected, player_clicks
    # Reset card
    card_picked = None
    # Reset valid moves
    valid_moves = []
    # Reset user clicks
    sq_selected = ()
    player_clicks = []


"""
Global dictionary of pieces and cards. Pictures and loaded only once on program start.
Source of cards: https://www.ultraboardgames.com/onitama/game-rules.php
Source of pieces: https://drive.google.com/drive/folders/1qH7IQj5lj7o3MQIb5TAZhsDr_5f9p8aq
"""
def load_images(gs):
    # Assigning pieces using dictionary PIECE_IMAGES["wP"] = pieces/wP.png
    pieces = ["wP", "wK", "bP", "bK"]
    for piece in pieces:
        # Scale image using resolution
        PIECE_IMAGES[piece] = pygame.transform.scale(pygame.image.load("pieces/" + piece + ".png"), (SQ_SIZE, SQ_SIZE))

    # Load all playing card images using dictionary CARD_IMAGES["wP"] = pieces/wP.png
    card_names = list(gs.all_cards.keys())
    for card in card_names:
        CARD_IMAGES[card] = pygame.transform.scale(pygame.image.load("cards/" + card + ".jpg"), CARD_SIZE)


"""
Responsible for all graphics changes within current game state. Drawing and refresh board, pieces, cards and labels.
"""
def draw_game_state(screen, gs, valid_moves, sq_selected, which_card, my_font):
    # Reset screen
    screen.fill(pygame.Color(BACKGROUND_COLOR))
    # Draw squares of board
    draw_board(screen)
    # Highlight
    highlight_square(screen, gs, valid_moves, sq_selected)
    # Draw pieces on the top of board
    draw_pieces(screen, gs.board)
    # Draw holders for cards
    draw_card_holders(screen, gs)
    # Selected card highlight
    draw_card_highlight(screen, gs, which_card)
    # Draw playing label
    draw_player_label(screen, gs, my_font)


"""
Draw squares of board on the left side of window
"""
def draw_board(screen):
    global colors
    colors = [pygame.Color(LIGHT_COLOR), pygame.Color(DARK_COLOR)]
    for r in range(DIMENSION):
        for c in range(DIMENSION):
            if (r, c) == (0, 2) or (r, c) == (4, 2):
                pygame.draw.rect(screen, KING_COLOR, pygame.Rect(c * SQ_SIZE, r * SQ_SIZE, SQ_SIZE, SQ_SIZE))
                continue
            color = colors[(r + c) % 2]
            pygame.draw.rect(screen, color, pygame.Rect(c * SQ_SIZE, r * SQ_SIZE, SQ_SIZE, SQ_SIZE))


"""
Highlight clicked square which piece is selected
"""
def highlight_square(screen, gs, valid_moves, sq_selected):
    if sq_selected != ():
        r, c = sq_selected
        # sq_selected is piece than can be moved
        if gs.board[r][c][0] == ('w' if gs.white_to_move else "b"):
            # Highlight
            s = pygame.Surface((SQ_SIZE, SQ_SIZE))
            # Transparency value (0 -> transparent, 255 -> opaque)
            s.set_alpha(100)
            s.fill(pygame.Color(SQ_HIGHLIGHT_ACTIVE))
            screen.blit(s, (c * SQ_SIZE, r * SQ_SIZE))

            # Highlight moves
            s.fill(pygame.Color(SQ_HIGHLIGHT_PASSIVE))
            for move in valid_moves:
                if move.start_row == r and move.start_col == c:
                    screen.blit(s, (move.end_col * SQ_SIZE, move.end_row * SQ_SIZE))


"""
Draw pieces on the top of board.
"""
def draw_pieces(screen, board):
    for r in range(DIMENSION):
        for c in range(DIMENSION):
            piece = board[r][c]
            # Not empty
            if piece != "--":
                screen.blit(PIECE_IMAGES[piece], pygame.Rect(c * SQ_SIZE, r * SQ_SIZE, SQ_SIZE, SQ_SIZE))


"""
Draw card holders with cards inside. Two cards for black, one spare card and two cards for white player.
"""
def draw_card_holders(screen, gs):
    for i, card in enumerate(gs.selected_cards):
        x, y, w, h = (CARD_POS[i][0], CARD_POS[i][1], CARD_SIZE[0], CARD_SIZE[1])
        # Rotate 180 degrees to face player
        #image = pygame.transform.rotate(CARD_IMAGES[card], 180) if not gs.white_to_move else CARD_IMAGES[card]
        image = CARD_IMAGES[card]
        screen.blit(image, pygame.Rect(x, y, w, h))
        pygame.draw.rect(screen, pygame.Color(BLACK), pygame.Rect(x, y, w, h), width=1)


"""
Possible highlight for selected card. Also highlight cards for the current player
"""
def draw_card_highlight(screen, gs, which_card):
    # Blue highlight
    if gs.curr_p == gs.player_name \
            and gs.win_white is None \
            and not gs.black_disconnected \
            and not gs.white_disconnected \
            and not gs.broken_conn:
        if gs.player_name == gs.black_name:
            x, y, w, h = get_rect_tuple(0)
            pygame.draw.rect(screen, pygame.Color(CARD_HIGHLIGHT_PASSIVE), pygame.Rect(x, y, w, h), width=3)
            x, y, w, h = get_rect_tuple(1)
            pygame.draw.rect(screen, pygame.Color(CARD_HIGHLIGHT_PASSIVE), pygame.Rect(x, y, w, h), width=3)
        elif gs.player_name == gs.white_name:
            x, y, w, h = get_rect_tuple(3)
            pygame.draw.rect(screen, pygame.Color(CARD_HIGHLIGHT_PASSIVE), pygame.Rect(x, y, w, h), width=3)
            x, y, w, h = get_rect_tuple(4)
            pygame.draw.rect(screen, pygame.Color(CARD_HIGHLIGHT_PASSIVE), pygame.Rect(x, y, w, h), width=3)

    if which_card is not None \
            and not gs.black_disconnected \
            and not gs.white_disconnected \
            and not gs.broken_conn:
        # Actively selected card
        i = gs.get_id_by_card(which_card)
        x, y, w, h = get_rect_tuple(i)
        pygame.draw.rect(screen, pygame.Color(CARD_HIGHLIGHT_ACTIVE), pygame.Rect(x, y, w, h), width=5)


"""
Return selected card for highlighting
"""
def get_rect_tuple(i):
    return CARD_POS[i][0], CARD_POS[i][1], CARD_SIZE[0], CARD_SIZE[1]


"""
Animate move of piece for one second
"""
def animate_move(move, screen, board, clock):
    global colors
    d_r = move.end_row - move.start_row
    d_c = move.end_col - move.start_col
    frames_per_square = 10
    frame_count = (abs(d_r) + abs(d_c)) * frames_per_square

    for frame in range(frame_count + 1):
        r, c = (move.start_row + d_r*frame/frame_count, move.start_col + d_c*frame/frame_count)
        draw_board(screen)
        draw_pieces(screen, board)

        # Erase
        color = colors[(move.end_row + move.end_col) % 2]
        end_square = pygame.Rect(move.end_col*SQ_SIZE, move.end_row*SQ_SIZE, SQ_SIZE, SQ_SIZE)
        pygame.draw.rect(screen, color, end_square)

        # Draw captured pieces
        if move.piece_captured != "--":
            screen.blit(PIECE_IMAGES[move.piece_captured], end_square)

        # Draw moving piece
        screen.blit(PIECE_IMAGES[move.piece_moved], pygame.Rect(c*SQ_SIZE, r*SQ_SIZE, SQ_SIZE, SQ_SIZE))
        pygame.display.flip()
        clock.tick(60)


"""
Draw label to inform which player is playing and some other info regarding connection and game over.
"""
def draw_player_label(screen, gs, my_font):
    black_who = "You" if gs.player_name == gs.black_name else "Enemy"
    white_who = "You" if gs.player_name == gs.white_name else "Enemy"

    black_status = "[PLAYING]" if not gs.white_to_move else ""
    white_status = "[PLAYING]" if gs.white_to_move else ""

    # Won / Lost
    if gs.win_white is not None:
        black_status = "[WON]" if not gs.win_white else "[LOST]"
        white_status = "[WON]" if gs.win_white else "[LOST]"

    # Disconnect
    elif gs.black_disconnected:
        black_status = "[DISCONNECT]"
        white_status = "[PAUSE]"
    elif gs.white_disconnected:
        black_status = "[PAUSE]"
        white_status = "[DISCONNECT]"

    # BROKEN CONN
    elif gs.broken_conn:
        black_status = "[BROKEN CONN]"
        white_status = "[BROKEN CONN]"

    black_name_lb = my_font.render(f"{black_who}: {gs.black_name}", True, BLACK)
    screen.blit(black_name_lb, BLACK_NAME_POS)

    black_status_lb = my_font.render(f"{black_status}", True, BLACK)
    screen.blit(black_status_lb, BLACK_STATUS_POS)

    spare_card_lb = my_font.render(f"Spare Card", True, BLACK)
    screen.blit(spare_card_lb, SPARE_CARD_POS)

    white_name_lb = my_font.render(f"{white_who}: {gs.white_name}", True, BLACK)
    screen.blit(white_name_lb, WHITE_NAME_POS)

    white_status_lb = my_font.render(f"{white_status}", True, BLACK)
    screen.blit(white_status_lb, WHITE_STATUS_POS)

